mono/sgen/sgen-alloc.c

```diff
@@ -246,6 +258,12 @@ sgen_alloc_obj_nolock (GCVTable vtable, size_t size)
 
 				zero_tlab_if_necessary (p, size);
 			} else {
+				/* Subtract size because the size of the object we are trying to allocate
+				 * will either be counted in the next TLAB,
+				 *  or in alloc_degraded.
+				 */
+				increment_thread_allocation_counter(TLAB_NEXT - TLAB_START - size);
```

Comment:
Is the temporary subtraction visible from outside and significant?

Reply:
I don't understand this line. Why are we decrementing size here ? Also we should bump the size for the previous tlab only after we are done allocating from it.

Reply:
My understanding of the code: If we are in this case, we have already moved  TLAB_NEXT beyond the end of the TLAB by size. We are therefore not going to allocate in this TLAB, and shouldn't count the allocation here. Thus the subtraction. 

This current allocation will be satisfied by either allocating in the next TLAB or using alloc_degraded. In the first case, it gets counted when *that* TLAB is retired. In the second, it is counted directly in alloc_degraded. Either way, it shouldn't be counted for the TLAB we are retiring now, or it else it will be double counted.

This is a little confusing, which is why I wrote the comment. I'm not sure of a more clear way to do it.

@jaykrell The subtraction is not visible from outside.

Reply:
1. We do TLAB_NEXT -= size above, so we are undo'ing this _invalid_ bump
2. Assume you do a degraded_alloc for this current size (and you already bumped the size for the entire tlab with this line) and then a smaller size is allocated next that fits in this still valid tlab. Wouldn't we then double count the tlab data ?
3. Now that I think about it, when doing a collection we will clear tlabs with allocated data that will not be accounted for (see sgen_clears_tlabs). We will need to bump data there.
4. This is getting a little bit complex. We should add a small test that allocates objects in a loop and checks that this counter returns valid data. Some considerations would be that on 32 bit the object header is 8 bytes, while on 64 bit it is 16 bytes, and all objects are 8 byte aligned (the allocated object size is multiple of 8, see SGEN_ALLOC_ALIGN)